clippy_lints/src/question_mark.rs

```diff
@@ -122,8 +153,56 @@ impl QuestionMark {
         }
     }
 
-    fn result_check_and_early_return(cx: &LateContext<'_>, expr: &Expr<'_>, nested_expr: &Expr<'_>) -> bool {
-        Self::is_result(cx, expr) && Self::expression_returns_unmodified_err(cx, nested_expr, expr)
+    /// Check whether the encountered expression contains `else if`
+    fn expr_is_else_if(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+        if_chain! {
+            if let ExprKind::If(..) = expr.kind;
+            if let Some(parent_expr) = clippy_utils::get_parent_expr(cx, expr);
+            if let ExprKind::If(..) = parent_expr.kind;
+            then {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Check whether the fixed code needs semicolon after `?`
+    ///
+    /// The fixed code of this expression doesn't need semi
+    /// ```ignore
+    /// let _ = if let Some(x) = Some(0) {
+    ///     x
+    /// } else {
+    ///     return None;
+    /// };
+    /// ```
+    ///
+    /// The fixed code of this expression needs semi
+    /// ```ignore
+    /// if let Err(err) = foo() {
+    ///     return Err(err);
+    /// }
+    /// ```
+    fn is_semicolon_needed(if_then: &Expr<'_>, if_else: Option<&'_ Expr<'_>>) -> bool {
+        let if_then_kind = &peel_blocks_with_stmt(if_then).kind;
+        let if_then_is_ret_stmt = matches!(if_then_kind, ExprKind::Ret(..));
+
+        if if_else.is_none() {
+            return if_then_is_ret_stmt;
+        }
+        if let ExprKind::Ret(_) = peel_blocks_with_stmt(if_else.unwrap()).kind {
+            return if_then_is_ret_stmt;
+        }
+        false
+    }
+
+    fn result_check_and_early_return(
+        cx: &LateContext<'_>,
+        expr: &Expr<'_>,
+        nested_expr: &Expr<'_>,
+        pat_symbol: Option<Symbol>,
+    ) -> bool {
+        Self::is_result(cx, expr) && Self::expression_returns_unmodified_err(cx, nested_expr, expr, pat_symbol)
```

Comment:
I think we could use some refactoring to avoid code repetition in a lot of these utils. For example, don't get the expression type in all the `is_*` functions. And just look for `ExprKind::Ret` in one place rather than having a bunch of `*_and_early_return` utils. Maybe you can have a refactor commit before other changes?

Reply:
I tried refactoring it, and it still looks like speghatti unfortunatly

Reply:
Instead of `is_early_return(..) || is_early_return(..)`, try doing the following. I think that will make things a lot simpler.

```rust
if let Some(adt) = caller_ty.ty_adt_ty();
if let Some(name)  = cx.tcx.get_diagnostic_name(adt.did);
if matches!(name, sym::Option | sym::Result);
```

Reply:
Ummm, I tried copy-paste this code, but I think it just brings back the issue #7859.